source/opt/local_single_block_elim_pass.h

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2017 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_LOCAL_SINGLE_BLOCK_ELIM_PASS_H_
+#define LIBSPIRV_OPT_LOCAL_SINGLE_BLOCK_ELIM_PASS_H_
+
+
+#include <algorithm>
+#include <map>
+#include <queue>
+#include <utility>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "basic_block.h"
+#include "def_use_manager.h"
+#include "module.h"
+#include "pass.h"
+
+namespace spvtools {
+namespace opt {
+
+// See optimizer.hpp for documentation.
+class LocalSingleBlockElimPass : public Pass {
+ public:
+  LocalSingleBlockElimPass();
+  const char* name() const override { return "eliminate-local-single-block"; }
+  Status Process(ir::Module*) override;
+
+ private:
+  // Returns true if |opcode| is a non-ptr access chain op
+  bool IsNonPtrAccessChain(const SpvOp opcode) const;
+
+  // Returns true if |typeInst| is a scalar type
+  // or a vector or matrix
+  bool IsMathType(const ir::Instruction* typeInst) const;
+
+  // Returns true if |typeInst| is a math type or a struct or array
+  // of a math type.
+  bool IsTargetType(const ir::Instruction* typeInst) const;
+
+  // Given a load or store |ip|, return the pointer instruction.
+  // Also return the base variable's id in |varId|.
+  ir::Instruction* GetPtr(ir::Instruction* ip, uint32_t* varId);
+
+  // Return true if |varId| is a previously identified target variable.
+  // Return false if |varId| is a previously identified non-target variable.
+  // If variable is not cached, return true if variable is a function scope 
+  // variable of target type, false otherwise. Updates caches of target 
+  // and non-target variables.
+  bool IsTargetVar(uint32_t varId);
+
+  // Replace all instances of |loadInst|'s id with |replId| and delete
+  // |loadInst|.
+  void ReplaceAndDeleteLoad(ir::Instruction* loadInst, uint32_t replId);
+
+  // Return true if any instruction loads from |varId|
+  bool HasLoads(uint32_t varId) const;
+
+  // Return true if |varId| is not a function variable or if it has
+  // a load
+  bool IsLiveVar(uint32_t varId) const;
+
+  // Return true if |storeInst| is not to function variable or if its
+  // base variable has a load
+  bool IsLiveStore(ir::Instruction* storeInst);
+
+  // Add stores using |ptr_id| to |insts|
+  void AddStores(uint32_t ptr_id, std::queue<ir::Instruction*>* insts);
+
+  // Delete |inst| and iterate DCE on all its operands 
+  void DCEInst(ir::Instruction* inst);
```

Comment:
The idea here was to derive some efficiency from the fact that we had some idea of what needed to be deleted, what likely needed to be deleted and what likely didn't need to be deleted.

For instance, for this pass, we knew that any loads whose ids were replaced could be deleted without having to look at use/def info. We also knew which loads' we didn't have to look at. Maybe not a huge efficiency gain, but something.

Yes, we do go back and (admittedly inefficiently) re-scan each block looking for stores  to delete. But we could have saved off the stores that we did link with a replaced load and only considered them and avoided a full scan. We still could do that and maybe we should.

My thinking on DCEInst was that it would be a generic recursive instruction deleter used by all the passes (again, possibly part of the proposed Optimizer Context), so the complexity would be somewhat limited in that way.

If we did write a separate DCE pass, we would probably want to design it to optionally only look at loads and stores, instead of all instructions. And we would need to do the load/store analysis in two full scans: first to remove any and all loads, and another scan to look at the stores, which can't be fully considered until all possible loads are removed.

While I think I still favor a more targeted approach to deal with load and store deletion, I am open to continuing to consider splitting this into a separate pass, so I will add it as a TODO.